hierarchicalforecast/core.py

```diff
@@ -90,170 +102,314 @@ class HierarchicalReconciliation:
+                )
+            if Y_hat_df_nw[model].is_null().any():
+                raise Exception(
+                    f"Column `{model}` in `Y_hat_df` contains null values. Make sure no column in `Y_hat_df` contains null values."
+                )
 
-        uids = Y_hat_df.index.unique()
+        # TODO: Complete y_hat_insample protection
+        model_names = [
+            name
+            for name in model_names
+            if not ("-lo" in name or "-hi" in name or "-median" in name)
+        ]
+        if intervals_method in ["bootstrap", "permbu"] and Y_df_nw is not None:
+            if not (set(model_names) <= set(Y_df_nw.columns)):
+                raise Exception(
+                    f"Check `Y_df` columns, {model_names} must be in `Y_df` columns."
+                )
+
+        # Assert S is an identity matrix at the bottom
+        S_df_nw_cols.remove(id_col)
+        if not np.allclose(
+            S_df_nw[S_df_nw_cols][-S_df_nw[S_df_nw_cols].shape[1] :],
+            np.eye(S_df_nw[S_df_nw_cols].shape[1]),
+        ):
+            raise ValueError(
+                f"The bottom {S_df_nw.shape[1]}x{S_df_nw.shape[1]} part of S must be an identity matrix."
+            )
 
         # Check Y_hat_df\S_df series difference
-        S_diff  = len(S_df.index.difference(uids))
-        Y_hat_diff = len(Y_hat_df.index.difference(S_df.index.unique()))
-        if S_diff > 0 or Y_hat_diff > 0:
-            raise Exception(f'Check `S_df`, `Y_hat_df` series difference, S\Y_hat={S_diff}, Y_hat\S={Y_hat_diff}')
-
-        if Y_df is not None:
-            # Check Y_hat_df\Y_df series difference
-            Y_diff  = len(Y_df.index.difference(uids))
-            Y_hat_diff = len(Y_hat_df.index.difference(Y_df.index.unique()))
-            if Y_diff > 0 or Y_hat_diff > 0:
-                raise Exception(f'Check `Y_hat_df`, `Y_df` series difference, Y_hat\Y={Y_hat_diff}, Y\Y_hat={Y_diff}')
-
-        # Same Y_hat_df/S_df/Y_df's unique_id order to prevent errors
-        S_df = S_df.loc[uids]
-
-        return Y_hat_df, S_df, Y_df, model_names
-
-    def reconcile(self, 
-                  Y_hat_df: pd.DataFrame,
-                  S: pd.DataFrame,
-                  tags: Dict[str, np.ndarray],
-                  Y_df: Optional[pd.DataFrame] = None,
-                  level: Optional[List[int]] = None,
-                  intervals_method: str = 'normality',
-                  num_samples: int = -1,
-                  seed: int = 0,
-                  sort_df: bool = True,
-                  is_balanced: bool = False,
-        ):
+        # TODO: this logic should be method specific
+        S_diff = set(S_df_nw[id_col]) - set(Y_hat_df_nw[id_col])
+        Y_hat_diff = set(Y_hat_df_nw[id_col]) - set(S_df_nw[id_col])
+        if S_diff:
+            raise ValueError(
+                f"There are unique_ids in S_df that are not in Y_hat_df: {S_diff}"
```

Comment:
These could be a lot, might be worth using reprlib, e.g.
```suggestion
                f"There are unique_ids in S_df that are not in Y_hat_df: {reprlib.repr(S_diff)}"
```
same for the ones below.